pkg/istio-agent/xds-proxy.go

```diff
@@ -0,0 +1,329 @@
+	}
+
+	return &oauth2.Token{
+		AccessToken: tok,
+		Expiry:      time.Now().Add(ts.period),
+	}, nil
+}
+
+func (p *XdsProxy) initDownstreamServer() error {
+	l, err := setUpUds("./etc/istio/proxy/XDS")
+	if err != nil {
+		return err
+	}
+	grpcs := grpc.NewServer()
+	discovery.RegisterAggregatedDiscoveryServiceServer(grpcs, p)
+	reflection.Register(grpcs)
+	p.downstreamGrpcServer = grpcs
+	p.downstreamListener = l
+	return nil
+}
+
+func buildUpstreamClientDialOpts(sa *Agent) ([]grpc.DialOption, error) {
+	tlsOpts, err := getTLSDialOption(sa)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build TLS dial option to talk to upstream: %v", err)
+	}
+
+	keepaliveOption := grpc.WithKeepaliveParams(keepalive.ClientParameters{
+		Time:    30 * time.Second,
+		Timeout: 10 * time.Second,
+	})
+
+	initialWindowSizeOption := grpc.WithInitialWindowSize(int32(defaultInitialWindowSize))
+	initialConnWindowSizeOption := grpc.WithInitialConnWindowSize(int32(defaultInitialConnWindowSize))
+	msgSizeOption := grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(defaultClientMaxReceiveMessageSize))
+	dialOptions := []grpc.DialOption{
+		tlsOpts,
+		grpc.WithConnectParams(grpc.ConnectParams{
+			Backoff:           backoff.DefaultConfig,
+			MinConnectTimeout: 5 * time.Second,
+		}),
+		keepaliveOption, initialWindowSizeOption, initialConnWindowSizeOption, msgSizeOption,
+	}
+
+	// TODO: This is not a valid way of detecting if we are on VM vs k8s
+	// Some end users do not use Istiod for CA but run on k8s with file mounted certs
+	// In these cases, while we fallback to mTLS to istiod using the provisioned certs
+	// it would be ideal to keep using token plus k8s ca certs for control plane communication
+	// as the intention behind provisioned certs on k8s pods is only for data plane comm.
+	if sa.secOpts.ProvCert == "" {
+		// only if running in k8s pod
+		dialOptions = append(dialOptions, grpc.WithPerRPCCredentials(oauth.TokenSource{&fileTokenSource{
+			"./var/run/secrets/tokens/istio-token",
+			time.Second * 300,
+		}}))
+	}
+	return dialOptions, nil
+}
+
+// Returns the TLS option to use when talking to Istiod
+// If provisioned cert is set, it will return a mTLS related config
+// Else it will return a one-way TLS related config with the assumption
+// that the consumer code will use tokens to authenticate the upstream.
+func getTLSDialOption(agent *Agent) (grpc.DialOption, error) {
+	var certPool *x509.CertPool
+	var err error
+	var rootCert []byte
+	if agent.secOpts.ProvCert != "" {
+		// This is most likely a VM using pre-provisioned mTLS certs (key.pem, cert-chain.pem, root-cert.pem)
+		// to talk to Istiod setup mtls
+		rootCert, err = ioutil.ReadFile(agent.secOpts.ProvCert + "/root-cert.pem")
+	} else if agent.secOpts.PilotCertProvider != "" {
+		rootCert, err = agent.loadPilotCertProviderRootCert()
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	certPool = x509.NewCertPool()
+	ok := certPool.AppendCertsFromPEM(rootCert)
```

Comment:
will this fail if `rootCert` is nil? I think its valid right?

Reply:
Yep. checked code